File: btagsil/repl.py
from sys import exit

import btagsil.data as data
import btagsil.world as w
from btagsil.world import World

# Immutable responses with info about the world

def repl_where(world: World, what: list[str]) -> str:
    match what:
        case []: return data.where_what()
        case ["is"]: return data.where_is_what()
        case ["is", *what]: return data.dont_know_where_is(' '.join(what))
        case ["am", "i"]: return w.current_loc_description(world)
        case ["can", "i", "go"]: return w.possible_destinations(world)
        case _: return data.where_error()

# Actions that mutate the world

def repl_go(world: World, where: list[str]) -> str:
    match where:
        case []: return data.can_only_go_to_stuff()
        case ["to"]: return data.go_to_where()
        case ["to", *where]: return w.move_to(world, where)
        case x: data.can_only_go_to_stuff()

def repl_look(world: World, what: list[str]) -> str:
    match what:
        case []: return data.empty_look()
        case ["at"]: return data.look_at_what()
        case ["at", *what]: return w.look_at_object(world, what)
        case ["around"]: return w.look_around(world)
        case _: return data.can_only_look_like_this()
        

# REPL actions

def repl_unknown(what: list[str]) -> None:
    print(data.unknown_action(' '.join(what)))

def repl_exit(args: list[str]) -> None:
    match args:
        case []:
            print(data.exit_message())
            exit(0)
        case x: print(data.exit_with_args(' '.join(x)))

def repl_echo(what: list[str]) -> None:
    match what:
        case []: print(data.echo_what())
        case x: print(f"{' '.join(x)}")

def repl_empty() -> None:
    """This is an empty action that does nothing. Occurs when user inputs nothing (or spaces) and presses enter."""

# Entry point for execution of a single action

def repl_once(world: World, input: str) -> None:
    tags = input.split()
    match tags:
        case []: repl_empty()
        case ["echo", *args]: repl_echo(args)
        case ["exit", *lol]: repl_exit(lol)
        case ["where", *what]: print(repl_where(world, what))
        case ["look", *what]: print(repl_look(world, what))
        case ["go", *where]: print(repl_go(world, where))
        case wtf: repl_unknown(wtf)
    print() if tags else print(end="", sep="")

File: src/main/scala/btagsil/Repl.scala
package btagsil

// Immutable nagivational and iformational actions

def actWhere(world: World, what: List[String]): String = what match {
    case List() => Text.whereWhat()
    case List("am", "i") => Info currentLocDescription world
    case List("can", "i", "go") => Info possibleDestinations world
    case List("is") => Text.whereIsWhat()
    case "is" :: thing => Text dontKnowWhereThingIs (thing mkString " ")
    case _ => Text.wrongWhere()
}

def actWhat(world: World, what: List[String]): String = what match {
    case List() => Text.whatWhat()
    case List("is") => Text.whatIsWhat()
    case "is" :: entity => Text.dontKnowWhatThingIs(entity mkString " ")
    case List("can", "i", "buy") => Info purchaseOptions world
    case _ => Text.wrongWhat()
}

def actLook(world: World, where: List[String]): String = where match {
    case List() => Text.tryLookingAtSomething()
    case List("around") => Info lookAround world
    case List("at") => Text.lookAtWhat()
    case List("at", "my", "weapon") => Text.youSeeYourWeapon(world.player.weapon.description)
    case "at" :: entity => Info.lookAtEntity(world, entity)
    case _ => Text.wrongLook()
}

def actTalk(world: World, target: List[String]): String = target match {
    case List() => Text.talkToSomeonePls()
    case List("to") => Text.talkToWho()
    case "to" :: entity => Info.talkToEntity(world, entity)
    case _ => Text.wrongTalk()
}


// Actions that mutate the world

def actGo(world: World, how: List[String]): String = how match {
    case List() => Text.canOnlyGoToPlaces()
    case List("to") => Text.go_to_where()
    case "to" :: where => Change.go_to_loc(world, where mkString " ")
    case _ => Text.canOnlyGoToPlaces()
}

def actBuy(world: World, what: List[String]): String = what match {
    case List() => Text.buyWhat()
    case thing :: "from" :: Nil => Text.buyFromWho(thing)
    case thing :: "from" :: seller => Change.buy(world, thing, seller)
    case _ => Text.needToBuyFromSomeone()
}

def actAttack(world: World, target: List[String]): String = target match {
    case List() => Text.smackWho()
    case entity => Change.smack(world, entity)
}


// REPL's system actions

def replRespond(response: String): Unit =
    println(response + "\n")

def replExit(): Unit =
    println(Text.exitMessage())
    System exit 0

def replEcho(what: List[String]): Unit =
    println(Text echoed (what mkString " "))

def replUnknown(wtf: List[String]): Unit =
    println(Text unknownInput (wtf mkString " "))

def replEmpty(): Unit = {}

def replOnce(world: World, action: List[String]): Unit = action match {
    case List() => replEmpty()
    case List("exit") => replExit()
    case "where" :: what => replRespond(actWhere(world, what))
    case "what" :: what => replRespond(actWhat(world, what))
    case "look" :: where => replRespond(actLook(world, where))
    case "talk" :: target => replRespond(actTalk(world, target))
    case "go" :: how => replRespond(actGo(world, how))
    case "buy" :: what => replRespond(actBuy(world, what))
    case "attack" :: target => replRespond(actAttack(world, target))
    case "echo" :: what => replEcho(what)

    case wtf => replUnknown(wtf)
}
